misc: Allow non-serializing behaviour for MiscRegs

giactra opened the pull request:
We traditionally enforce System register read/writes consistency by marking writes as:

`IsNonSpeculative && IsSerializeAfter`

and marking reads as:

`IsSerializeBefore`

For example in Arm this means the instructions following the MSR (write) will block in the rename stage
until the MSR gets executed and committed, while the MRS (read) will block itself in rename until
the ROB is empty

This is somehow necessary because MiscReg writes are buffered in the execution context (the DynInst in o3)
and the architectural regfile (in the ISA object) only gets updated when the MSR gets committed

While this could make sense for particular system registers with side effects, some registers can
safely be updated speculatively. Imposing a serializing behaviour for every misc reg read/write is a big
performance penalty especially for cores with a big instruction window.
Being able to read and write without serializing behaviour means allowing the CPU to recognise
producer-consumer dependencies between system register writes/reads

This PR is adding the capability in the O3 CPU of handling non-serializing misc reg updates and it does
so by starting to track aforementioned dependencies through the scoreboard. Any ISA could use this
logic by marking their MiscRegs as non-serializing.

**Note1**: This still requires the system register write to be non-speculative as we currently don't implement rollback
of system register write in case of mispeculation

**Note2**: This PR does not implement renaming for MiscRegs. This means WAW dependencies are handled in
the rename stage by effectively tagging a non-serializing MiscReg write as serializing in case there is already a
pending producer.

Commit:
arch-arm: Add serializing flag to MISCREGS

We extend the MiscRegInfo enum with the MISCREG_SERIALIZING
flag, to mark Misc registers as serializing/non-serializing
on a per-register basis.
We then generate different read/write instructions accordingly.
with different StaticInstFlags

By default a MiscReg is marked as serializing.

We expect to slowly transition most of the sysregs into the
non-serializing subset

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I86973780f3b67850b5ea0e6a530f698fb175d148

Commit:
cpu: Define a serializing flag for RegClasses

This patch adds a serializing flag to RegClasses.

At the moment (and by default) this is effectively
identical to the isFixedMapping method: it calls
the isRenameable method which returns false for
InvalidRegClass and MiscRegClass registers.

So the meaning of the method is: can the following RegId
be read/written without serializing behaviour?

The reason why we add this "duplicate" is because we
want to change the behaviour for some MiscRegClass
registers. While the registers are generally speaking
non-renameable, we still want to provide the capability
to some of those to be read/written without serializing
execution, provided the register is properly tracked in
the scoreboard

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: Ia53494208ec6e674ca4a433a1e40795ecd8e675b

Commit:
cpu: Add isAlwaysReady method to the PhysRegId

Some physical registers do not use the scoreboard and
are always marked as being ready.

They mainly rely on the fact there can never be a direct
producer/consumer relationship between two instructions.

Writes to those registers are only executed at commit
time (IsNonSpeculative) and they stall any younger
instructions in rename (IsSerializeAfter) until the
write has committed and the arch regfile has been
updated

Reads are also stalled in rename (IsSerializeBefore) until
the ROB is empty to make sure every possible producer
has written back the results to the arch regfile

With the patch we stop treating fixed mapping registers
as always being ready. Having a fixed mapping shouldn't
mean the register is intrinsically ready.
We add a isAlwaysReady method which checks if the
register can be updated speculatively. This means there
has to be a dependency tracking mechanism at play so we
shouldn't assume the register is always ready

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I0df9af2a949dd8d4cf8b8cd40e8a3308f42bcca4

Commit:
cpu-o3: Allow non-serializing update of MiscRegClass registers

With this commit we finally track some MiscRegs with the
o3 Scoreboard, allowing some of the to be speculatively
updated. This needs to be enabled on the ISA side
(see incoming patches)

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I00aa08b414364123becb5d2faf8eba3da902934a

Commit:
arch-arm: Allow non-serializing updates of TPIDR_ELx register

The TPIDR_ELx registers are scratch registers software
can read/write to.

As the PE makes no use of those registers, they
can be read/written speculatively to improve performance
without serializing execution

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: Ifc5f22f253ff4e020859d13982aaafa971935832

Commit:
cpu-o3: Add matching DPRINTF to the o3 Scoreboard

We are printing whenever a scoreboard entry is marked
as ready through we are not doing the same when
it is marked as busy.

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I91a5cf963cf8dc3dcc8eeb1a8e219df1556a9608

Commit:
arch-arm: Treat MSR/MRS to SP_ELx differently

This patch is treating read/writes to SP_ELx registers
(SP_EL0, SP_EL1, SP_EL2) as special cases of MSR/MRS.

We traditionally enforce System register read/writes
consistency by marking writes as:

IsNonSpeculative && IsSerializeAfter [1]

and marking reads as:

IsSerializeBefore [2]

This means the instructions following the MSR will
block in the rename stage until the MSR gets executed
and committed, while the MRS will block itself in
rename until the ROB is empty

This is somehow necessary because MiscReg writes are
buffered in the execution context (the DynInst in o3) [3]
and the architectural regfile (in the ISA object)
only gets updated when the MSR gets committed

While this could make sense for particular system
registers with side effects, some registers can
safely be updated speculatively

[1]: https://github.com/gem5/gem5/blob/stable/\
    src/arch/arm/isa/insts/data64.isa#L356
[2]: https://github.com/gem5/gem5/blob/stable/\
    src/arch/arm/isa/insts/data64.isa#L333
[3]: https://github.com/gem5/gem5/blob/stable/
    src/cpu/o3/dyn_inst.hh#L1029
[4]: https://github.com/gem5/gem5/blob/stable/\
    src/cpu/o3/dyn_inst.hh#L1070

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: Ia04c88779987dc6315c8d6099ceead62214d094f

Commit:
arch-arm: Allow non-serializing reads of ID registers

The are RO registers, whose value cannot be modified throughout
execution. There is no risk in marking them as
non-serializing regs so that the MRS can be renamed without
waiting for the ROB to be empty (IsSerializeBefore)

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I23c5f16362c99da4c267d2bb086f7c2fe9ca88d0

Commit:
cpu, configs: Add a SystemOp OpClass

Change-Id: I6a716f90690fb765938f584b415f951156da1340
Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>

Commit:
arch-arm: Tag MSR/MRS instructions as System instructions

Change-Id: I23bf81e5ab4e241eea685233b218fe40172e066c
Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>

Commit:
misc: Remove unused IprAccess OpClass

This was used by the alpha ISA and would be the analogous
of the newly added SystemOp class

Change-Id: Ibb94ec56cdab57e6f8306cd5630c98c16ff2b429
Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>

Commit:
arch: Make the ControlRegOperand inherit from RegOperand

Because it's not inheriting from it, we are missing the
_numTypedDestRegs array write for the MiscReg case.

This means that numDestRegs(MiscRegClass) will always
return 0;

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I20610e8ff9d9195acca9198ad2373d0a7d4d1baa

dhschall commented on src/arch/x86/X86CPU.py:
```diff
@@ -80,7 +80,6 @@ class DefaultX86FUPool(FUPool):
         PredALU(),
         WritePort(),
         RdWrPort(),
-        IprPort(),
```
I think this should be replaced with `System_Unit`

giactra replied:
Done

dhschall commented on src/cpu/o3/dyn_inst.hh:
```diff
@@ -1066,6 +1066,23 @@ class DynInst : public ExecContext, public RefCounted
+            assert(isNonSpeculative());
+
+            bool no_squash_from_TC = thread->noSquashFromTC;
+            thread->noSquashFromTC = true;
```
Why is that needed?
Will the `setMisc` function do something weird?

giactra replied:
Copy pasted from the updateMiscRegs logic:

```
1092         // @todo: Pretty convoluted way to avoid squashing from happening when
1093         // using the TC during an instruction's execution (specifically for
1094         // instructions that have side-effects that use the TC).  Fix this.
1095         // See cpu/o3/dyn_inst_impl.hh.
```

IIRC there are cases where calling

thread->setMiscReg would cause a squash in the O3 model due to the following:

https://github.com/gem5/gem5/blob/stable/src/cpu/o3/thread_context.cc#L207C16-L207C26

The idea is that if you are changing anything through the TC interface, it means you are forcing a change in the execution context and therefore you need to squash the pipeline.

powerjg commented:
At a high-level this all sounds good to me. I don't know if I'll get a chance to review carefully in the next ~week or so, though. Let's just make sure that Linux boots with RISC-V and x86 before merging.

dhschall commented on src/arch/arm/regs/misc.cc:
```diff
@@ -4826,81 +4842,97 @@ ISA::initializeMiscRegMetadata()
       .allPrivileges().exceptUserMode().writes(0);
     InitReg(MISCREG_ID_PFR0_EL1)
       .allPrivileges().exceptUserMode().writes(0)
+      .serializing(false)
```
What determines an instruction that does not need to be serialized?

dhschall reviewed (commented):
Thank you @giactra !
This is a huge help for us. Overall it looks good to me and while testing it we saw a significant reduction in serialization stalls without any bug.

My main question is, in the `misc.cc` file you specify manually which instructions can be serialized or not. How do you know which are does not need to be serialized?
Is it possible that other instructions can be marked as non-serializing which are not yet marked?


### Our Analysis

@yongjiehuang, a student of mine did extensive tests with our suite of server workloads which have a lot of accesses to MISC registers.

Here are the results. 
<img width="719" height="260" alt="image" src="https://github.com/user-attachments/assets/6e5f3acc-922a-41b3-9470-e870560ff891" />

Your patch indeed reduced the stalls for most benchmarks. However, there are still a few that have a lot. He analyzed which instructions caused the stalls for the two outlier benchmarks ([The patch to trace serializing instructions](https://github.com/dhschall/gem5-fdp/commit/b475d23b7719c10d994656835bec3f4ed2ad0839))

Below the results for `otmetrics` (the other benchmark looks the same)
<img width="480" height="239" alt="image" src="https://github.com/user-attachments/assets/86b50053-7539-4219-855d-ba4a8633de20" />

Most of the stalls are still initiated by MRS/MSR instructions. Before it was mostly reads from special registers now its writes.
In particular, writes to `daif` and `spsr_el1`. So the question is how do we know if these registers can also be marked as non-serializing without trying?

Another question is why do we still see `sp_el0` in the trace? The stack pointer is now mapped to an integer register and should not need to be serialized or not?

powerjg commented:
I believe at least one of the CI test failures are real. It looks like `build/VEGA_X86/gem5.opt` is failing to build. (I could be wrong, the tests have been a little flaky lately.)

The failure on the fast unittests also seems real. It ended with the same error twice. Plus, it looks like it's related to the changed code: 

```
/usr/bin/ld: build/ALL/arch/arm/aapcs64.test.tfo: in function `gem5::ArmISA::MiscRegClassOps::serializing(gem5::RegId const&) const':
aapcs64.test.cc:(.text._ZNK4gem56ArmISA15MiscRegClassOps11serializingERKNS_5RegIdE[_ZNK4gem56ArmISA15MiscRegClassOps11serializingERKNS_5RegIdE]+0x12): undefined reference to `gem5::ArmISA::lookUpMiscReg'
```

giactra commented:
> I believe at least one of the CI test failures are real. It looks like `build/VEGA_X86/gem5.opt` is failing to build. (I could be wrong, the tests have been a little flaky lately.)
> 
> The failure on the fast unittests also seems real. It ended with the same error twice. Plus, it looks like it's related to the changed code:
> 
> ```
> /usr/bin/ld: build/ALL/arch/arm/aapcs64.test.tfo: in function `gem5::ArmISA::MiscRegClassOps::serializing(gem5::RegId const&) const':
> aapcs64.test.cc:(.text._ZNK4gem56ArmISA15MiscRegClassOps11serializingERKNS_5RegIdE[_ZNK4gem56ArmISA15MiscRegClassOps11serializingERKNS_5RegIdE]+0x12): undefined reference to `gem5::ArmISA::lookUpMiscReg'
> ```

Yes I am trying to fix that. The issue is the unit-tests try to include as little as possible. By referencing the lookUpMiscReg in the misc.hh header file I am forced to add misc.cc to the source list

Which I'd rather avoid as it will bring lots of code inside the binary.. So I am trying to restructure code so that we don't end up including misc.cc

powerjg commented:
> Which I'd rather avoid as it will bring lots of code inside the binary.. So I am trying to restructure code so that we don't end up including misc.cc

Yes, please! As you know, I've been on the crusade recently as well 😄

giactra commented:
> Thank you @giactra ! This is a huge help for us. Overall it looks good to me and while testing it we saw a significant reduction in serialization stalls without any bug.
> 
> My main question is, in the `misc.cc` file you specify manually which instructions can be serialized or not. How do you know which are does not need to be serialized? Is it possible that other instructions can be marked as non-serializing which are not yet marked?

100%

It is not specified in the PR message per se but It is in one of the commits: the current labelling of system register is not exhaustive:
```


By default a MiscReg is marked as serializing.
 
We expect to slowly transition most of the sysregs into the
non-serializing subset

```

There are many other register which could be marked as non-serializing. For instance, the CONTEXTIDR register.
In general every register with no side-effect can be easily transitioned into the non-serializing bunch. For registers with side effects, it is a bit more complex.

A register with side effects can only be transitioned if we are able to establish dependencies (in the o3) between the register write and any instruction which depends on its value

giactra commented:
> Another question is why do we still see `sp_el0` in the trace? The stack pointer is now mapped to an integer register and should not need to be serialized or not?

Thanks for the report. I don't know for sure, but looking at your patch, I believe the issue might not be the MRS SP_EL0 per se, but rather the preceding instruction.

In fact SerializeAfter does nothing more than labelling the next instruction in PO as SerializeBefore.

So if you have

MSR SCTLR_EL1 x0 -> SerializeAfter
MRS x1, SP_EL0 

The SP read will be tagged as SerializeBefore by the preceding instruction. So in your evaluation you need to distinguish between static and temporary SerializeBefore. If you find some time to debug your experiments, let me know your finding

dhschall commented:
> In fact SerializeAfter does nothing more than labelling the next instruction in PO as SerializeBefore.

Thanks! that might be indeed correct. I'll ask the student if he can distinguish between instructions that are marked by previous instructions

dhschall commented on src/cpu/o3/rename.cc:
```diff
@@ -1155,6 +1170,38 @@ Rename::renameDestRegs(const DynInstPtr &inst, ThreadID tid)
 }
 
+void
+Rename::setSerializeBefore(DynInstPtr &inst, ThreadID tid)
```
Tbh. I found this a bit misleading. The function suggests you set serialization regardless. However, what the function actually does it check for WaW. If one exists, resolve it with serialization.
I think `checkMiscWaWConflict` or something similar would be more appropriate

giactra replied:
I agree, I thought about it. I will rename accordingly

dhschall commented:
Just to be clear what do you mean by _"side-effect"_
For example we are now trying `SPSR_EL1` and `ELR_EL1`.
SPSR_EL1: _Holds the saved process state when an exception is taken to EL1._
Would you call it side-effect free?

Commit:
arch-arm: Move MiscReg funcs from utility to misc

The utility.hh header contains multiple functions used throughout the
ArmISA code. These don't necessarily have anything in common;
some of them should also be located in more appropriate files.

As such utility.hh is heavily included and includes many
other header files itself.

To cleanup the utility.hh header, we move MiscReg specific
helpers (like the decoding/faulting ones) to regs/misc.hh
which is a more natural place for them.

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: Iaf1f030fec6d1c1132862e1ecb24a18062a34ad7

Commit:
arch-arm: Move MMU::CachedState to source file

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I3faab444ba618df43bd596d2c8586bf0da049d19

Commit:
arch-arm: Do not include self_debug.hh in ISA header file

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I67710634087e83e468d602fa326b99498404c98a

Commit:
arch-arm: Define a new misc_info.hh file

This new MiscReg specific couple of source/header file
is supposed to hold informations about MiscRegs. It is
supposed to be more hidden than the misc.hh file, which
contains the definition of MiscRegIndexes and therefore
is the interface to system registers

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: I7047a93da2961ea7e115488dc5518d942e3dcea6

Commit:
arch-arm: Move definition of MiscReg related methods to isa.cc

Signed-off-by: Giacomo Travaglini <giacomo.travaglini.arm.com>
Change-Id: Ia2e64ee62a98ac64584618a3127392b3371f3d9e

giactra commented:
> Just to be clear what do you mean by _"side-effect"_ For example we are now trying `SPSR_EL1` and `ELR_EL1`. SPSR_EL1: _Holds the saved process state when an exception is taken to EL1._ Would you call it side-effect free?

One **initial** definition of side effects (in our case, because for the Arm architecture is a bit different) could be:

> A register whose value will be consumed by an instruction different from an explicit read

For instance a MSR TPIDR_EL0 does not have side effects because apart for a MRS TPIDR_EL0 no other instruction will try to read that value. So as long as you keep track of the dependency via the scoreboard, the scheduler/O3 engine will ensure the RAW dependency is preserved.

The SPSR_EL1 value is actually consumed in several circumstances, namely when you ERET and when you throw an exception.

However, considering those are serializing-non speculative events, I suppose you should be able to mark the SPSR as non serializing. Unless I am missing something.

powerjg commented on src/proto/inst.proto:
```diff
@@ -95,7 +95,6 @@ message Inst {
     SIMDFloatSqrt = 29;
     MemRead = 30;
     MemWrite = 31;
-    IprAccess = 32;
```
To ensure we don't accidentally reuse the field number 32 in the future, it's best practice in Protobuf to explicitly reserve it after deletion. This prevents potential bugs with older clients that might still be sending or expecting the old IprAccess field.

Could we please add the following reserved keywords?

```Protocol Buffers
    reserved 32;
    reserved "IprAccess";
```

This will safeguard our protocol's backward compatibility. Thanks!

giactra replied:
Ack

powerjg reviewed (commented):
lgtm. Please take a look at David's suggestions/questions as well. One small thing on the protobuf change.
